keps/sig-autoscaling/20181106-in-place-update-of-pod-resources.md

```diff
@@ -0,0 +1,376 @@
+
+Setting the flag to separately control CPU & memory is due to an observation
+that usually CPU can be added/removed without much problem whereas changes to
+available memory are more probable to require restarts.
+
+If more than one resource type with different policies are updated, then
+RestartPod policy takes precedence over RestartContainer, which in turn takes
+precedence over NoRestart policy.
+
+#### CRI Changes
+
+Kubelet calls UpdateContainerResources CRI API which currently takes
+*runtimeapi.LinuxContainerResources* parameter that works for Docker and Kata,
+but not for Windows. This parameter is changed to *runtimeapi.ContainerResources*,
+that is runtime agnostic.
+
+### Flow Control
+
+When a new Pod is created, Scheduler is responsible for selecting a suitable
+Node that accommodates the Pod. When a Pod resize is requested, Kubelet attempts
+to update the Pod and its Containers. If required, Kubelet may evict lower
+priority Pods from the Node to make room for the resize. Scheduler observes the
+resize request, and updates the accounting in its cache.
+
+The following steps denote a typical flow of an in-place resize operation for a Pod
+with ResizePolicy set to NoRestart for all its Containers.
+
+1. Initiating actor updates Pod's Container.ResourceRequirements using PATCH verb.
+1. API Server validates the new ResourceRequirements (e.g. limits are not below
+   requested resources, QoS class does not change, ResourceQuota not exceeded..).
+1. API Server calls all Admission Controllers to verify the Pod Update.
+   * If any of the Controllers reject the update, API Server responds with an
+     appropriate error message.
+1. API Server updates PodSpec object with the new desired ResourceRequirements.
+1. Kubelet observes that Pod's Container.ResourceRequirements and
+   ContainerStatus.ResourcesAllocated differ. It checks its Node allocatable
+   resources to determine if the new desired ResourceRequirements fit the Node.
+   * _Case 1_: Kubelet finds new ResourceRequirements fit. It sets
+     ResizingPod.Status to true, and ResizingPod.Reason to InProgress, then
+     applies resized cgroup limits to the Pod and its Containers, and once
+     successfully done, updates Pod's ContainerStatus.ResourcesAllocated to
+     reflect the desired ResourceRequirements. It then sets ResizingPod.Status
+     to false, and ResizingPod.Reason to Success.
+     - If at the same time, a new Pod was scheduled against the capacity used by
+       this resource resize, that Pod is rejected during Kubelet admission if
+       Node has no more room.
+   * _Case 2_: Kubelet finds that desired ResourceRequirements does not fit.
+     - Kubelet checks to see if evicting lower priority Pods on the Node can
+       successfully resize the Pod. If yes, it sets ResizingPod.Status to true,
+       ResizingPod.Reason to PreEmpting, and initiates pre-emption of lower
+       priority Pods via Eviction API. Once lower priority Pods have been
```

Comment:
There are a few cases to consider here noting that all involved components evaluate used capacity with max(Resources,ResourcesAllocated):
a) Scheduler uses max(Resources,ResourcesAllocated) it should not assign any new Pods to this node after it sees the resize. A new Pod that was assigned in parallel to Kubelet handling the resize will be rejected if we are at capacity during admission control based on Kubelet using max(Resources,ResourcesAllocated)
b) If I'm not mistaken, HandlePodUpdates invoked through the channel seralizes the resize requests. And using max(Resources,ResourcesAllocated) allows Kubelet to accept the first one that gets there, and fail subsequent resizes if we are at capacity.

If there is capacity, both new Pod add and resize updates will go through as it does today.

Did I miss something?

Reply:
See my comments above, as I'm still confused about the implementation of retries for admission.  If there are multiple resizes that have been "rejected" by the node, and are awaiting retry, how do we make sure the "right" (e.g. highest priority) pod is the pod allowed to run after another pod is evicted?

Reply:
Perhaps the following example will help illustrate:

1. T=0: Let's say a node has 4G mem, and has 4 pods with following mem requests: P1a (1G), P1b (1G), P2 (1G), P3 (1G) with priorities P1a == P1b > P2 > P3.
2. T=1: Let's say P1a wants to resize to 3G, and HandlePodUpdates determines it is possible if it evicts P2, P3, and invokes eviction API (which say takes 10s for P2, 30s for P3), then sets P1a ResizingPod condition to 'InProgress' with timestamp=1, and returns.
3. T=x(1-11) For the duration of eviction of P2, sum[P1a (max(1G,3G)), P1b (1G), P2 (1G), P3 (1G)] = 6G. Any new Pod scheduled in parallel (race-condition) is rejected at KL admission as 6G > 4G, so no capacity.
4. T=2: P1b wants to resize to 1.01G, but it will go to 'FailedNodeCapacity' becasue sum[P1a (max(1G,3G)), P1b max(1G,1.01G), P2 (1G), P3 (1G)] = 6.01G > 4G, so no capacity.
5. T=11: P2 is gone, and now sum[P1a (max(1G,3G)), P1b (max(1G,1.01G)), P3 (1G)] = 5.01G. Event based or periodic retry for P1b resize will continue to stay at 'FailedNodeCapacity' because 5.01G > 4G. Scheduler should not schedule any new pods due to the same max(Resources, ResourcesAllocated) logic.
6. T=31: P3 is gone. P1a moves from 'InProgress' to 'Success', ResourcesAllocated=3G. P1b retry continues to result in 'FailedNodeCapacity' condition because sum[P1a (3G), P1b (max(1G,1.01G))] = 4.01G > 4G.

If P1b was higher priority than P1a, we start eviction of P1a at T=2.
If in the above scenario, P1a was resizing to 2.99G instead of 3G, then P1b would get resized at T=31 retry.

Does that help?

Reply:
@dashpole When you have the chance, could you please review the above example and lmk if this helps clarify the flow? thanks.

Reply:
Yeah, that clarifies it for me.  There is a race condition here, but we are saying that _eventually_ the high-priority pod will be given those resources.  It just might take a lot of evictions for it to get to increase in size if there are many other pods waiting to resize.  That is probably the best we can do without scheduler involvement or implementing a priority-queue for admission.  We should make a note that this is intentional.

Re-asking, since it wasn't answered: Does the eviction API guarantee the pod will be removed in a specific time period?

Reply:
I really don't know if eviction API guarantees pod removal in a certain amount of time, the [documentation](https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/#the-eviction-api) does not call it out. I'll try it out later tonight and see what I find out..

Reply:
I tried out a few simple cases calling eviction API and works as expected - if PDB is available and eviction is granted, we get back code 201, and the pod is deleted. If not, we get back error 429, and we need to go down the retry path. Are there any cases you're aware of where the eviction API can cause problems?

Reply:
As @derekwaynecarr mentioned at our last meeting, we should ideally be able to establish SLOs around resizes, at least for requests.  I was wondering if the pod termination grace period is respected by the eviction API, as that could make evictions take an unlimited amount of time, and prevent us from placing an upper bound on how long.  For kubelet evictions, we don't give the pod its grace period for this reason, as we want the resources immediately.

@bsalamat when the scheduler preempts a pod from a node, does it respect graceful termination periods?  We should try and maintain the same approximate behavior here.

Reply:
Eviction is not part of this KEP now, resolving this issue.